From 0280a659b7489df3af1a296f77643673c6b9c9f3 Mon Sep 17 00:00:00 2001
From: YouMin Chen <cym@rock-chips.com>
Date: Mon, 17 Apr 2023 11:30:52 +0800
Subject: [PATCH] PM / devfreq: rockchip_dmc: Add lock for
 rockchip_dmcfreq_get_dev_status

RK3588 platform may crash if the CPU and MCU access the DFI/DMC
registers at same time.

Signed-off-by: YouMin Chen <cym@rock-chips.com>
Change-Id: I6912f72aff09f7db6d696f4a5b15ccc06f68ba04
---
 drivers/devfreq/rockchip_dmc.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

--- a/drivers/devfreq/rockchip_dmc.c
+++ b/drivers/devfreq/rockchip_dmc.c
@@ -598,12 +598,17 @@ static int rockchip_dmcfreq_get_dev_stat
 	if (!dmcfreq->info.auto_freq_en)
 		return -EINVAL;
 
+	/*
+	 * RK3588 platform may crash if the CPU and MCU access the DFI/DMC
+	 * registers at same time.
+	 */
+	rockchip_monitor_volt_adjust_lock(dmcfreq->mdev_info);
 	for (i = 0; i < dmcfreq->edev_count; i++) {
 		ret = devfreq_event_get_event(dmcfreq->edev[i], &edata);
 		if (ret < 0) {
 			dev_err(dev, "failed to get event %s\n",
 				dmcfreq->edev[i]->desc->name);
-			return ret;
+			goto out;
 		}
 		if (i == dmcfreq->dfi_id) {
 			stat->busy_time = edata.load_count;
@@ -613,7 +618,10 @@ static int rockchip_dmcfreq_get_dev_stat
 		}
 	}
 
-	return 0;
+out:
+	rockchip_monitor_volt_adjust_unlock(dmcfreq->mdev_info);
+
+	return ret;
 }
 
 static int rockchip_dmcfreq_get_cur_freq(struct device *dev,
